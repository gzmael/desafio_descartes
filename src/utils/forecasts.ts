import { Forecast } from 'models/Forecast'

export const getDayByDate = (date: string) =>
  new Intl.DateTimeFormat('pt-br', {
    day: '2-digit',
    timeZone: 'America/Fortaleza',
  }).format(new Date(date))

export const getHourByTimestamp = (time: number) =>
  new Intl.DateTimeFormat('pt-br', {
    hour: '2-digit',
    minute: '2-digit',
    timeZone: 'America/Fortaleza',
  }).format(new Date((time + 10800) * 1000))

export const getDayWeekByDate = (
  date: string,
  style?: 'short' | 'long' | 'narrow',
) =>
  new Intl.DateTimeFormat('pt-br', {
    weekday: style ?? 'long',
    timeZone: 'America/Fortaleza',
  }).format(new Date(date))

export const getWeekAndDaysFromForecasts = (forecasts: Forecast[]) => {
  const dates = [
    ...new Set(forecasts.map((forecast) => forecast.date.split(' ')[0])),
  ]
  return dates.map((date) => {
    return {
      date,
      day: getDayByDate(date + ' 12:00:00'),
      week: getDayWeekByDate(date + ' 12:00:00', 'short'),
    }
  })
}
