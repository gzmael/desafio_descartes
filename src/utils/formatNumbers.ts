export const formatTemperature = (temp: number, maxFrac = 1) => {
  return new Intl.NumberFormat('pt-br', {
    maximumFractionDigits: maxFrac,
  }).format(temp)
}

export const formatMsToKmH = (value: number, digits = 2) => {
  return new Intl.NumberFormat('pt-br', {
    maximumFractionDigits: digits,
  }).format(value * 3.6)
}
