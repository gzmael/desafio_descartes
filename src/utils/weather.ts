import { Weather, WeatherResponse } from 'models/Weathear'
import { formatMsToKmH, formatTemperature } from './formatNumbers'
import { Forecast, ForecastReponse } from 'models/Forecast'

export const getDetailListFromWeather = (weather: Weather) => {
  return [
    {
      title: 'Vento',
      value: formatMsToKmH(weather.wind) + ' km/h',
    },
    {
      title: 'Umidade',
      value: weather.humidity + ' %',
    },
    {
      title: 'Sensação',
      value: formatTemperature(weather.feels) + ' ºC',
    },
    {
      title: 'Min - Máx',
      value: `${formatTemperature(weather.min, 0)}ºC - ${formatTemperature(
        weather.max,
        0,
      )}ºC`,
    },
    {
      title: 'Visibilidade',
      value: weather.visibility / 1000 + ' km',
    },
    {
      title: 'Nuvens',
      value: weather.clouds + ' %',
    },
  ]
}

export const getWeatherFromData = (data: WeatherResponse) => {
  const { clouds, main, visibility, weather, wind, sys } = data
  return {
    clouds: clouds.all,
    description: weather[0].description,
    feels: main.feels_like,
    humidity: main.humidity,
    icon: weather[0].icon,
    max: main.temp_max,
    min: main.temp_min,
    temperature: main.temp,
    wind: wind.speed,
    pressure: main.pressure,
    sunrise: sys.sunrise,
    sunset: sys.sunset,
    visibility,
  }
}

export const getForecastsFromData = (data: ForecastReponse): Forecast[] => {
  return data.list.map((forecast) => {
    const {
      clouds,
      dt_txt: date,
      main,
      visibility,
      weather,
      wind,
      dt,
    } = forecast
    return {
      clouds: clouds.all,
      description: weather[0].description,
      feels: main.feels_like,
      humidity: main.humidity,
      icon: weather[0].icon,
      max: main.temp_max,
      min: main.temp_min,
      temperature: main.temp,
      wind: wind.speed,
      pressure: main.pressure,
      visibility,
      date,
      timestamp: dt,
    }
  })
}
