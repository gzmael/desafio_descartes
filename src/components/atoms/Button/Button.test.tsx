import { render, screen } from '@testing-library/react'

import { Button, ButtonProps, defaultStyle } from '.'
import { cn } from 'utils/cls-merge'

describe('Component/Atom/Button', () => {
  it('should render Button with default variant', () => {
    const props: ButtonProps = {
      children: 'Button Default',
    }

    const classNames = cn([
      defaultStyle,
      'h-10 pl-3 pr-4 gap-2 text-base btn-md',
      'w-auto',
      'bg-indigo-500 hover:bg-indigo-600 text-white hover:text-white focus:ring-2 focus:ring-blue-400 focus:ring-offset-2',
      'rounded',
    ])

    const { container } = render(<Button {...props} />)

    expect(screen.getByText(/Button Default/i)).toBeInTheDocument()

    expect(container.firstChild).toHaveClass(classNames)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render Button with icon variant', () => {
    const props: ButtonProps = {
      children: 'Button Icon',
      color: 'icon',
    }

    const classNames = cn([
      defaultStyle,
      'h-10 pl-3 pr-4 gap-2 text-base btn-md',
      'w-auto',
      'bg-transparent hover:bg-transparent transition-all focus:ring-none',
      'rounded',
      'p-2',
    ])

    const { container } = render(<Button {...props} />)

    expect(screen.getByText(/Button Icon/i)).toBeInTheDocument()

    expect(container.firstChild).toHaveClass(classNames)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render small primary Button with variants', () => {
    const props: ButtonProps = {
      children: 'Button Primary',
      color: 'primary',
      size: 'sm',
      radii: 'none',
      w: 'half',
    }

    const classNames = cn([
      defaultStyle,
      'h-8 pl-2 pr-3 gap-1 text-sm btn-sm',
      'w-auto',
      'bg-indigo-500 hover:bg-indigo-600 text-white hover:text-white focus:ring-2 focus:ring-blue-400 focus:ring-offset-2',
      'rounded-none',
      'w-3/4',
    ])

    const { container } = render(<Button {...props} />)

    expect(screen.getByText(/Button Primary/i)).toBeInTheDocument()

    expect(container.firstChild).toHaveClass(classNames)

    expect(container.firstChild).toMatchSnapshot()
  })
})
