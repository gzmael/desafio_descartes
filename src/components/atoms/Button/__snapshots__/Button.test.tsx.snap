// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component/Atom/Button > should render Button with default variant 1`] = `
<button
  class="flex items-center justify-center font-medium disabled:cursor-not-allowed cursor-pointer transition-all disabled:opacity-50 disabled:pointer-events-none focus:outline-none active:scale-95 bg-indigo-500 hover:bg-indigo-600 text-white hover:text-white focus:ring-2 focus:ring-blue-400 focus:ring-offset-2 h-10 pl-3 pr-4 gap-2 text-base btn-md rounded w-auto"
>
  Button Default
</button>
`;

exports[`Component/Atom/Button > should render Button with icon variant 1`] = `
<button
  class="flex items-center justify-center font-medium disabled:cursor-not-allowed cursor-pointer disabled:opacity-50 disabled:pointer-events-none focus:outline-none active:scale-95 bg-transparent hover:bg-transparent transition-all focus:ring-none h-10 gap-2 text-base btn-md rounded w-auto p-2"
>
  Button Icon
</button>
`;

exports[`Component/Atom/Button > should render small primary Button with variants 1`] = `
<button
  class="flex items-center justify-center font-medium disabled:cursor-not-allowed cursor-pointer transition-all disabled:opacity-50 disabled:pointer-events-none focus:outline-none active:scale-95 bg-indigo-500 hover:bg-indigo-600 text-white hover:text-white focus:ring-2 focus:ring-blue-400 focus:ring-offset-2 h-8 pl-2 pr-3 gap-1 text-sm btn-sm rounded-none w-3/4"
>
  Button Primary
</button>
`;
